chalk-solve/src/clauses.rs

```diff
@@ -67,44 +109,45 @@ pub fn push_auto_trait_impls<I: Interner>(
         1
     );
 
+    // we assume that the builder has no binders so far.
+    assert!(builder.placeholders_in_scope().is_empty());
+
+    // auto traits are not implemented for foreign types
+    if let TypeName::Foreign(_) = app_ty.name {
```

Comment:
Hmm, I feel like it would be simpler to make `constituent_types` return an `Option` and then have it return `None` for this.